src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -202,6 +212,67 @@ private bool TryExtractStringConcatArgs(BoundExpression lowered, out ImmutableAr
                         return true;
                     }
 
+                    break;
+
+                case BoundSequence { SideEffects.Length: 0, Value: BoundCall sequenceCall } sequence:
+                    var locals = sequence.Locals;
+
+                    if ((object)sequenceCall.Method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_2ReadOnlySpans) ||
+                        (object)sequenceCall.Method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_3ReadOnlySpans) ||
+                        (object)sequenceCall.Method == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__Concat_4ReadOnlySpans))
+                    {
+                        // Faced a span-based `string.Concat` call. Since we can produce such call on the previous iterations ourselves, we need to unwrap it.
+                        // The key thing is that we need not to only extract arguments, but also unwrap them from being spans and for chars also wrap them into `ToString` calls.
+                        var wrappedArgs = sequenceCall.Arguments;
+                        var unwrappedArgsBuilder = ArrayBuilder<BoundExpression>.GetInstance(capacity: wrappedArgs.Length);
+
+                        for (var i = 0; i < wrappedArgs.Length; i++)
+                        {
+                            var wrappedArg = wrappedArgs[i];
+
+                            // Check whether a call is an implicit `string -> ReadOnlySpan<char>` conversion
+                            if (wrappedArg is BoundCall { Method: var argMethod, Arguments: [var singleArgument] } &&
+                                (object)argMethod == _compilation.GetWellKnownTypeMember(WellKnownMember.System_String__op_Implicit_ToReadOnlySpanOfChar))
+                            {
+                                unwrappedArgsBuilder.Add(singleArgument);
+                            }
+                            // This complicated check is for a sequence, which wraps a span around single char.
+                            // The sequence needs to have this shape: `{ locals: <none>, sideEffects: temp = <original char expression>, result: new ReadOnlySpan<char>(in temp) }`
+                            else if (wrappedArg is BoundSequence
+                            {
+                                Locals.Length: 0,
+                                SideEffects: [BoundAssignmentOperator { Right.Type.SpecialType: SpecialType.System_Char } assignment],
+                                Value: BoundObjectCreationExpression { Constructor: var objectCreationConstructor, Arguments: [BoundLocal constructorLocal] }
```

Comment:
>objectCreationConstructor [](http://example.com/codeflow?start=88&length=25)

Is it Ok for this value to be null?

Reply:
Bound nodes are null-annotated and `Constructor` property of `BoundObjectCreationExpression` is not marked as nullable (and there is [an assert in the constructor](https://github.com/dotnet/roslyn/blob/4cd098b66619b74098103d75e6de18484272de99/src/Compilers/CSharp/Portable/Generated/BoundNodes.xml.Generated.cs#L6244) to verify that)